### ISSUE TITLE
[BN][Navi21] `#undef __AMDGCN__` means UB

### ISSUE THREAD

**Issue by atamazov** [ROCmSoftwarePlatform/MIOpen#1036]
`#undef __...` should NOT be used in kernels.

----
This change leads to undefined behavior. We shouldn't undef macros predefined by the compiler.

_Originally posted by @atamazov in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/493#r671445106_

**Comment by atamazov** [comment]
Another place: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/00d57548c25399186eb201eb2e0e6f7480915c48/src/kernels/batchnorm_functions.h#L134-L136

**Comment by junliume** [comment]
> `#undef __...` should NOT be used in kernels.
> 
> This change leads to undefined behavior. We shouldn't undef macros predefined by compiler.
> 
> _Originally posted by @atamazov in [#493 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/493#r671445106)_

Do we have a ticket tracking this issue?

**Comment by junliume** [comment]
> Another place:
> 
> https://github.com/ROCmSoftwarePlatform/MIOpen/blob/00d57548c25399186eb201eb2e0e6f7480915c48/src/kernels/batchnorm_functions.h#L134-L136

Do push_macro and pop_macro work?

```
// file head
#pragma push_macro("__AMDGCN__")
#undef __AMDGCN__

// content

//file end
#pragma pop_macro("__AMDGCN__")
```

Yikes, it is already in a header file and doubled in many files mentioned above. 

**we need some clean up work on Batch Norm kernel code**, suggestion:
1. If every batch norm kernels have batchnorm_functions.h included at the very top of file, and all of them **need** this macro from compiler to be undefined: then generate another restore_batchnorm_macros.h and include them at the bottom of all of the files. **prone to errors, ugly, and some personal dislikes**

2. Move it out of header file, I think it should not be there in the first place. Do the push and pop in individual files if you have to.

**Comment by junliume** [comment]
@muralinr could you help to check if the above may work? Or with Artem for a full blown solution? Thanks!

**Comment by muralinr** [comment]
@junliume Sure Jun.  Let me work with Artem.

**Comment by atamazov** [comment]
> Do push_macro and pop_macro work?

It may work (and most likely will) but this is UB anyway, because there is #undef still.

Recommendation:
```C++
#ifdef __AMDGCN__
#define MIOPEN_USE_AMDGCN 1
#endif
...
if <something>
#undef MIOPEN_USE_AMDGCN
#define MIOPEN_USE_AMDGCN 0
#endif
...
// now use "#if MIOPEN_USE_AMDGCN" instead of "#ifdef __AMDGCN__".
```

**Comment by atamazov** [comment]
via #1037
